aws_lambda_powertools/utilities/idempotency/persistence/redis.py

```diff
@@ -0,0 +1,440 @@
+        )
+
+        client = redis.Redis(
+            host="localhost",
+            port="6379",
+            decode_responses=True,
+        )
+        persistence_layer = RedisCachePersistenceLayer(client=client)
+
+        @idempotent(persistence_store=persistence_layer)
+        def lambda_handler(event: dict, context: LambdaContext):
+            print("expensive operation")
+            return {
+                "payment_id": 12345,
+                "message": "success",
+                "statusCode": 200,
+            }
+        ```
+        """
+
+        # Initialize Redis client with Redis config if no client is passed in
+        if client is None:
+            self.client = RedisConnection(
+                host=host,
+                port=port,
+                username=username,
+                password=password,
+                db_index=db_index,
+                url=url,
+                mode=mode,
+                ssl=ssl,
+                ssl_cert_reqs=ssl_cert_reqs,
+            )._init_client()
+        else:
+            self.client = client
+
+        self.in_progress_expiry_attr = in_progress_expiry_attr
+        self.expiry_attr = expiry_attr
+        self.status_attr = status_attr
+        self.data_attr = data_attr
+        self.validation_key_attr = validation_key_attr
+        self._json_serializer = json.dumps
+        self._json_deserializer = json.loads
+        super(RedisCachePersistenceLayer, self).__init__()
+        self._orphan_lock_timeout = min(10, self.expires_after_seconds)
+
+    def _get_expiry_second(self, expiry_timestamp: int | None = None) -> int:
+        """
+        return seconds of timedelta from now to the given unix timestamp
+        """
+        if expiry_timestamp:
+            return expiry_timestamp - int(datetime.datetime.now().timestamp())
+        return self.expires_after_seconds
+
+    def _item_to_data_record(self, idempotency_key: str, item: Dict[str, Any]) -> DataRecord:
+        in_progress_expiry_timestamp = item.get(self.in_progress_expiry_attr)
+        if isinstance(in_progress_expiry_timestamp, str):
+            in_progress_expiry_timestamp = int(in_progress_expiry_timestamp)
+        return DataRecord(
+            idempotency_key=idempotency_key,
+            status=item[self.status_attr],
+            in_progress_expiry_timestamp=in_progress_expiry_timestamp,
+            response_data=str(item.get(self.data_attr)),
+            payload_hash=str(item.get(self.validation_key_attr)),
+            expiry_timestamp=item.get("expiration", None),
+        )
+
+    def _get_record(self, idempotency_key) -> DataRecord:
+        # See: https://redis.io/commands/get/
+        response = self.client.get(idempotency_key)
+        # key not found
+        if not response:
+            raise IdempotencyItemNotFoundError
+        try:
+            item = self._json_deserializer(response)
+        except json.JSONDecodeError:
+            raise IdempotencyOrphanRecordError
```

Comment:
Yes, I'll add it. Json decoding error is considered an Orphan Record cause this scenario will also introduce possible race condition just like Orphaned. As two lambda function is possible to reach this line of code almost simultaneously. They will both start the overwrite process without knowing each other. Causing this value being overwritten twice. Behaving exactly the same as Orphan Record does. So this case can also be handled by the error handling in `except IdempotencyOrphanRecordError` part to avoid the possible race condition.